components/calendar/calendar.js

```diff
@@ -551,6 +581,23 @@ class Calendar extends LocalizeCoreElement(RtlMixin(LitElement)) {
 		});
 	}
 
+	willUpdate(changedProperties) {
+		super.willUpdate(changedProperties);
+		if (changedProperties.get('_shownYear') === undefined && changedProperties.get('_shownMonth') === undefined) return;
+
+		const dates = getDatesInMonthArray(this._shownMonth, this._shownYear);
+
+		/** Dispatched when the calender view changes. "e.detail" provides the year and month in view. */
+		this.dispatchEvent(new CustomEvent('d2l-calendar-view-change', {
```

Comment:
Another way I had pictured this potentially working was to fire this event, but then expect consumers to attach the days that have events on them to the event detail. That would likely have to happen `async`, so maybe using some kind of Promise. This code would then wait, grab that and assign it to a state property which would trigger a re-render.

Pros and cons to each approach, and a little weirdness around the "first render" since it's a bit strange to fire an event before the user has even done anything.

Reply:
Yeah I am open to either approach. An alternate approach we talked about last week was to allow the user to pass serialized data as an attribute, which one or two other controls do. Both Margaree and I preferred the data component since passing data structures like that is sorta awkward. I know that's not exactly what you are proposing though.

Reply:
Right. I think the main challenge consumers is going to have is that there could be many, many events -- especially with recurring events -- so getting "all" the events isn't feasible. Instead, the API needs to support only fetching the events for the month (which your approach handles) or maybe even simpler only fetching the _days_ within the month that have events on them. I think that's technically all you need to render the dots? An API that includes the actual dates & time ranges is more flexible in case we want to change the UI later, but theoretically I think you just needs the days with true/false for each one.

Reply:
Yeah, for the moment we only need to know true/false for any given day, but we're definitely trying to leave our options open. Using consumer-specified colors (like Calendar color), displaying event info, and event counts have been mentioned but are out of scope for now. So the current approach is for them to tell us about the events, and we'll digest them to render dots or tooltips or whatever.

It makes me wonder, what if they wanted to display an indicator if an assignment is due that day? Is that an "event"? Should we be thinking something more generic?

Reply:
I was thinking about this a bunch overnight. The `start-value` + `end-value` was motivated by my initial look at the Calendar tool, considering a case like `Standard First Aid, Sat to Sun` as one event, but I now see there are other cases where the days are not consecutive. For example, `Standard First Aid, every Sat & Sun for 2 weeks`. So I feel like we need to give them an API with a single value property.
```html
<d2l-calendar-event value="2024-09-21">
<d2l-calendar-event value="2024-09-22">
<d2l-calendar-event value="2024-09-28">
<d2l-calendar-event value="2024-09-29">
```

Reply:
That's a good realization. Which... at that point since we've lost the potential connections between the different events (the first and last in your example _could_ be the same recurring event), is there much advantage to even calling them events vs. just an API that given a particular month (date range) returns the days in which one or more events occur?

Reply:
Yeah, it doesn't seem right to call them events. I think this could be done a couple different ways.

1. With a component like `d2l-calendar-day-info`. Due to `d2l-calendar` handling all the rendering (like `d2l-filter`), opening up date-specific `slot`s (for example for tooltip content) will never really be on the table, so HTML isn't buying us anything for that. So a component for specifying the data doesn't provide much value other than it being conventional.  In order for the consumer to know the exact date range in view, they would need to call `getDatesInMonthArray(...)` (which is already available) themselves, or we given them a new helper that gives them the min/max dates in view.

2. With a property `<d2l-calendar days-info="...">` + an event to update it when the view changes. Like the component approach above, this handles the initial state nicely (in theory no re-renders), but like option 1 above, the consumer would need call one of our helpers to get the min/max dates in view to provide the initial event data. No slot shenanigans.

3. With an event the consumer responds to when the component initially renders, and when the view changes. (this is the idea you mentioned earlier) Unless consumers query for event data in advance (again using the helpers mentioned in options 1 & 2), `d2l-calendar` would render at least once and later re-render once event data is fetched. The advantage of this approach is that, if the potential delay is acceptable, the consumer doesn't need to call the helpers to determine the min/max dates in view - they would be provided as part of the event detail.

Reply:
Yeah. Looks like when `selected-value` isn't provided, the initial month that gets displayed is... a complex combination of the current date, `max-value` and `min-value`. So yeah... the consumer isn't going to know the initial month unless they provide it themselves -- making a call to something like `getDatesInMonthArray` necessary like you said.

For 3, "with an event the consumer responds to when the component initially renders" sounds like it could be kind of flaky timing wise? I know we've had issues with events firing at component boot-up being missed by ancestor consumers.

Maybe an alternate version of 2 & 3 could leverage either [Lit contexts](https://lit.dev/docs/data/context/) or our own provider mixin to request an optional "calendar event info context" that the calendar could then call whenever it wanted to get the days with events? That would give us a lot of flexibility to make that context API whatever we want in the future. One thing I'm not sure about though is whether either of those mechanisms support the context being optional, which we'd obviously need.

Reply:
The `getDatesInMonthArray` expects the consumer to pass in the year and month for which they want dates. If we wanted to give them the ability to query in advance, we could factor out [that logic you mentioned](https://github.com/BrightspaceUI/core/blob/main/components/calendar/calendar.js#L587) to provide a helper like `getMinMaxDatesInView(selectedValue, minValue, maxValue)` that simply returns `{ minValue: ..., maxValue: ... }`. 

Yeah, I always worry a bit about event timing, but I think if consumers wire-up in the correct life-cycle method, then it works out ok. Handlers attaching data to event objects always seems a bit weird to me, but maybe if the event was named something like `d2l-calendar-info-request` it would seem more fitting. Our `requestInstance(...)`/`provideInstance(...)` abstracts that away and makes the code clearer imo. In either case, I think the setup for the consumer is pretty minimal.

If we request the event data in `willUpdate()`, we can pass along the dates in view, and if they happen to have the event data loaded then it would all happen in a single render cycle.

The Lit Context is interesting. Their examples are in TS so it's less clear what a vanilla setup looks like, but it looks like a more refined version of Justin's proposal that motivated our provider/requestor stuff. It seems like it might be a bit more involved for the consumer, but I'm still poking around.

Reply:
For the timing thing, I was specifically thinking about MVC-land where event listeners get wired up in class initializers, which will be well after Lit things have initialized and fired events.

Reply:
Maybe all the more reason to support specifying the initial event data via slotted elements or the attribute/property rather than relying exclusively on the event.